File: app/src/main/java/com/saienko/androidthings/barman/service/CocktailService.java
package com.saienko.androidthings.barman.service;

import android.app.IntentService;
import android.content.Context;
import android.content.Intent;
import android.util.Log;
import com.google.android.things.pio.Gpio;
import com.google.android.things.pio.PeripheralManagerService;
import com.saienko.androidthings.barman.Utils;
import com.saienko.androidthings.barman.db.cocktail.Cocktail;
import com.saienko.androidthings.barman.db.cocktail.CocktailElement;
import com.saienko.androidthings.barman.db.motor.Motor;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class CocktailService extends IntentService {

    private static final String TAG = "CocktailService";

    private static final String ACTION_COCKTAIL_CREATE = "ACTION_COCKTAIL_CREATE";
    private static final String ACTION_COCKTAIL_CANCEL = "ACTION_COCKTAIL_CANCEL";
    private static final String ACTION_MOTOR_TEST      = "ACTION_MOTOR_TEST";
    private static final String ACTION_MOTOR_CLEAR     = "ACTION_MOTOR_CLEAR";

    private static final String EXTRA_COCKTAIL   = "EXTRA_COCKTAIL";
    private static final String EXTRA_MOTOR      = "EXTRA_MOTOR";
    private static final String EXTRA_MOTOR_LIST = "EXTRA_MOTOR_LIST";

    public static final String BROADCAST_COCKTAIL_PROGRESS = "BROADCAST_COCKTAIL_PROGRESS";

    public static final String BROADCAST_COCKTAIL_STATUS         = "BROADCAST_COCKTAIL_STATUS";
    public static final String BROADCAST_EXTRA_COCKTAIL_STATUS   = "BROADCAST_EXTRA_COCKTAIL_STATUS";
    public static final String BROADCAST_EXTRA_COCKTAIL_ITEM_MAP = "BROADCAST_EXTRA_COCKTAIL_ITEM_MAP";

    private static boolean isTaskRun           = false;
    private static boolean taskShouldBeStopped = false;

    private HashMap<Long, Integer> progressMap;
    private ArrayList<Thread>      threadList;

    public CocktailService() {
        super("CocktailService");
    }

    public static void startCocktail(Context context, Cocktail cocktail) {
        Intent intent = new Intent(context, CocktailService.class);
        intent.setAction(ACTION_COCKTAIL_CREATE);
        intent.putExtra(EXTRA_COCKTAIL, cocktail);
        context.startService(intent);
    }

    public static void stopCocktail(Context context) {
        Intent intent = new Intent(context, CocktailService.class);
        intent.setAction(ACTION_COCKTAIL_CANCEL);
        context.startService(intent);
    }

    public static void motorTest(Context context, Motor motor) {
        Intent intent = new Intent(context, CocktailService.class);
        intent.setAction(ACTION_MOTOR_TEST);
        intent.putExtra(EXTRA_MOTOR, motor);
        context.startService(intent);
    }


    public static void clearMotors(Context context, List<Motor> motorList) {
        Intent intent = new Intent(context, CocktailService.class);
        intent.setAction(ACTION_MOTOR_CLEAR);
        intent.putParcelableArrayListExtra(EXTRA_MOTOR_LIST, (ArrayList<Motor>) motorList);
        context.startService(intent);
    }

    @Override
    protected void onHandleIntent(Intent intent) {
        if (intent != null) {
            final String action = intent.getAction();
            if (ACTION_COCKTAIL_CREATE.equals(action)) {
                final Cocktail cocktail = intent.getParcelableExtra(EXTRA_COCKTAIL);
                handleActionCocktail(cocktail);
            } else if (ACTION_COCKTAIL_CANCEL.equals(action)) {
                Log.d(TAG, "onHandleIntent: cancel");
                sendCocktailStatus(CocktailStatus.CANCEL);
                stopTask();
                Thread.currentThread().interrupt();
                stopSelf();
            } else if (ACTION_MOTOR_TEST.equals(action)) {
                motorTest(intent.getParcelableExtra(EXTRA_MOTOR));
                Log.i(TAG, "onHandleIntent: motorTest");
            } else if (ACTION_MOTOR_CLEAR.equals(action)) {
                clear(intent.getParcelableArrayListExtra(EXTRA_MOTOR_LIST));
            } else {
                throw new UnsupportedOperationException("Unknown action");
            }
        }
    }

    private void stopTask() {
        if (isTaskRun) {
            taskShouldBeStopped = true;
        }

        if (threadList != null) {
            for (Thread thread : threadList) {
                thread.interrupt();
            }
        }
    }

    private void handleActionCocktail(Cocktail cocktail) {
        taskShouldBeStopped = false;
        isTaskRun = true;
        progressMap = new HashMap<>();
        threadList = new ArrayList<>();
        for (CocktailElement cocktailElement : cocktail.getCocktailElements()) {
            progressMap.put(cocktailElement.getPosition().getMotor().getGpioId(), 0);
            pour(cocktailElement.getPosition().getMotor(), cocktailElement.getVolume());
        }
    }

    private void sendCocktailItemProgress(long gpioId, int progress) {
        progressMap.put(gpioId, progress);

        Intent intentUpdate = new Intent();
        intentUpdate.setAction(BROADCAST_COCKTAIL_PROGRESS);
        intentUpdate.addCategory(Intent.CATEGORY_DEFAULT);
        intentUpdate.putExtra(BROADCAST_EXTRA_COCKTAIL_ITEM_MAP, progressMap);
        sendBroadcast(intentUpdate);
    }

    private void sendCocktailStatus(CocktailStatus status) {
        Intent intentUpdate = new Intent();
        intentUpdate.setAction(BROADCAST_COCKTAIL_STATUS);
        intentUpdate.addCategory(Intent.CATEGORY_DEFAULT);
        intentUpdate.putExtra(BROADCAST_EXTRA_COCKTAIL_STATUS, status);
        sendBroadcast(intentUpdate);
    }

    private void motorTest(Motor motor) {
        pour(motor, 100);
    }

    private static void setLedValue(Gpio ledGpio, boolean value) {
        try {
            ledGpio.setValue(value);
        } catch (IOException e) {
            Log.e(TAG, "Error updating GPIO value", e);
        }
    }

    private void clear(List<Motor> motorList) {
        for (Motor motor : motorList) {
            if (Utils.isThingsDevice(this)) {
                pour(motor, 500);
            }
        }
    }

    private void startRealPourThread(int millis, Motor motor) {
        Thread thread = new Thread(() -> {

            Log.d(TAG, "startRealPourThread: before sleep GPIO" + motor.getGpio());

            long startTime = System.nanoTime();
            try {
                Thread.sleep(millis);
            } catch (InterruptedException e) {
                Log.e(TAG, "startRealPourThread: ", e);
                Thread.currentThread().interrupt();
            }
            long endTime  = System.nanoTime();
            long duration = (endTime - startTime);
            Log.d(TAG, "startRealPourThread: time " + duration + " GPIO " + motor.getGpio());
        });
        threadList.add(thread);
        thread.start();
    }

    private void startProgressThread(int step, Motor motor) {
        Thread thread = new Thread(() -> {
            Log.d(TAG, "startProgressThread: before sleep GPIO " + motor.getGpio());

            long startTime = System.nanoTime();
            for (int i = 0; i < 101; i++) {
                try {
                    Thread.sleep(step);
                    if (taskShouldBeStopped) {
                        Thread.currentThread().interrupt();
                        break;
                    }
                } catch (InterruptedException e) {
                    Log.e(TAG, "pourDemo: ", e);
                    Thread.currentThread().interrupt();
                }
                sendCocktailItemProgress(motor.getGpioId(), i);
            }
            long endTime  = System.nanoTime();
            long duration = (endTime - startTime);
            Log.d(TAG, "startProgressThread: time " + duration + " GPIO " + motor.getGpio());
        });
        thread.start();
    }

    private void pour(Motor motor, int ml) {
        Log.d(TAG, "pour() called with: motor = [" + motor + "], ml = [" + ml + "]");
        if (taskShouldBeStopped) {
            return;
        }
        Log.d(TAG, "pour() called with: motor = [" + motor + "], ml = [" + ml + "]");
        PeripheralManagerService pioService = null;
        if (Utils.isThingsDevice(this)) {
            pioService = new PeripheralManagerService();
        }
        Gpio mLedGpio = null;
        int  millis   = ml * motor.getMotorSpeed();
        int  step     = millis / 100;

        Log.d(TAG, "pour: millis = [" + millis + "], step= [" + step + "]");

        try {

            if (Utils.isThingsDevice(this)) {
                mLedGpio = pioService.openGpio(motor.getGpio().getGpioName());
                mLedGpio.setDirection(Gpio.DIRECTION_OUT_INITIALLY_LOW);
                setLedValue(mLedGpio, true);
            }
            Log.d(TAG, "pour: start sleep");
            startProgressThread(step, motor);
            startRealPourThread(millis, motor);

            if (Utils.isThingsDevice(this)) {
                setLedValue(mLedGpio, false);
            }

        } catch (IOException e) {
            Log.e(TAG, "Error configuring GPIO pins", e);
        }

        if (mLedGpio != null) {
            try {
                mLedGpio.close();
            } catch (IOException e) {
                Log.e(TAG, "Error closing LED GPIO", e);
            }
        }
    }
}

File: app/src/main/java/com/saienko/androidthings/barman/service/CocktailService.kt
package com.saienko.androidthings.barman.service

import android.app.IntentService
import android.content.Context
import android.content.Intent
import android.util.Log
import com.google.android.things.pio.Gpio
import com.google.android.things.pio.PeripheralManager
import com.saienko.androidthings.barman.R.string.motor
import com.saienko.androidthings.barman.Utils
import com.saienko.androidthings.barman.db.entity.Motor
import com.saienko.androidthings.barman.db.manager.CocktailManager
import com.saienko.androidthings.barman.db.manager.GpioManager
import kotlinx.coroutines.experimental.Job
import kotlinx.coroutines.experimental.cancelAndJoin
import kotlinx.coroutines.experimental.delay
import kotlinx.coroutines.experimental.launch
import java.io.IOException
import java.util.*
import java.util.concurrent.TimeUnit
import kotlin.collections.ArrayList

class CocktailService : IntentService(TAG) {

    private var progressMap: HashMap<Long, Int>? = null
    private var threadList: ArrayList<Job> = ArrayList()
    private var gpioList: ArrayList<Gpio> = ArrayList()


    override fun onHandleIntent(intent: Intent?) {
        if (intent != null && intent.action != null) {
            val action = intent.action
            when (action) {
                ACTION_COCKTAIL_CREATE -> {
                    val cocktailId = intent.getLongExtra(EXTRA_COCKTAIL_ID, -1)
                    handleActionCocktail(cocktailId)
                }
                ACTION_COCKTAIL_CANCEL -> {
                    Log.d(TAG, "onHandleIntent: cancel")
                    sendCocktailStatus(CocktailStatus.CANCEL)
                    launch { stopTask() }
                    Thread.currentThread().interrupt()
                    stopSelf()
                }
                ACTION_MOTOR_TEST -> {
                    val motor: Motor = intent.getParcelableExtra(EXTRA_MOTOR)
                    val motorSpeed = intent.getIntExtra(EXTRA_GPIO_SPEED, 1)
                    val gpioManager = GpioManager()
                    pour(motor.gpioId, gpioManager.get(motor.gpioId).gpioName, motorSpeed, 10, 1.0)
                    Log.i(TAG, "onHandleIntent: motorTest")
                }
                ACTION_MOTOR_CLEAR -> {
                    val motorList: List<Motor> = intent.getParcelableArrayListExtra(EXTRA_MOTOR_LIST)
                    val clearType = intent.getSerializableExtra(EXTRA_MOTOR_CLEAR_TYPE) as CleanType
                    val gpioManager = GpioManager()
                    motorList.forEach {
                        if (it.gpio == null) {
                            it.gpio = gpioManager.get(it.gpioId)
                        }
                    }
                    clear(motorList, clearType)
                }
                else -> throw UnsupportedOperationException("Unknown action")
            }
        } else {
            throw UnsupportedOperationException("Unknown action")
        }
    }

    private suspend fun stopTask() {
        if (isTaskRun) {
            taskShouldBeStopped = true
        }

        gpioList.forEach {
            it.close()
        }

        for (job in threadList) {
            job.cancelAndJoin()
        }
        threadList = ArrayList()
    }

    private fun handleActionCocktail(cocktailId: Long) {
        taskShouldBeStopped = false
        isTaskRun = true
        val cocktailManager = CocktailManager()

        val cocktail = cocktailManager.get(cocktailId)
        if (cocktail.cocktailElements != null) {
            for (cocktailElement in cocktail.cocktailElements!!) {
//                if (cocktailElement.position != null && cocktailElement.position!!.motor != null) {
//                    putToMap(cocktailElement.position!!.motor!!.gpioId, 0)
//                    cocktailElement.position!!.motor?.let {
//                        pour(it, cocktailElement.volume, cocktailElement.component.coefficient)
//                    }
//                }
                val component = cocktailElement.component
                cocktailElement.position?.let {
                    //                    if (component == null){
//                        val componentManager = ComponentManager()
//                        component = componentManager.get()
//                    }
                    putToMap(it.motor!!.gpioId, 0)
                    it.motor?.let {
                        pour(it, cocktailElement.volume, component?.coefficient ?: 1.0)
                    }
                }
            }
        }
    }

    private fun putToMap(gpioId: Long, progress: Int) {
        if (progressMap == null) {
            progressMap = HashMap()
        }
        progressMap!![gpioId] = progress
    }

    private fun sendCocktailItemProgress(gpioId: Long, progress: Int) {
        putToMap(gpioId, progress)

        val intentUpdate = Intent()
        intentUpdate.action = BROADCAST_COCKTAIL_PROGRESS
        intentUpdate.addCategory(Intent.CATEGORY_DEFAULT)
        intentUpdate.putExtra(BROADCAST_EXTRA_COCKTAIL_ITEM_MAP, progressMap)
        sendBroadcast(intentUpdate)
    }

    private fun sendCocktailStatus(status: CocktailStatus) {
        val intentUpdate = Intent()
        intentUpdate.action = BROADCAST_COCKTAIL_STATUS
        intentUpdate.addCategory(Intent.CATEGORY_DEFAULT)
        intentUpdate.putExtra(BROADCAST_EXTRA_COCKTAIL_STATUS, status)
        sendBroadcast(intentUpdate)
    }

    private fun clear(motorList: List<Motor>, clearType: CleanType) {
        Log.d(TAG, "clear motors: $motorList as ${clearType.name}")
        val volume: Int = when (clearType) {
            CleanType.LONG -> 500
            CleanType.SHORT -> 10
        }
        for (motor in motorList) {
            pour(motor, volume, 1.0)
        }
    }

    private fun startRealPourThread(millis: Int, gpioName: String) {
        lateinit var gpio: Gpio
        val job = launch {
            val manager: PeripheralManager
            if (Utils.isThingsDevice(applicationContext)) {
                manager = PeripheralManager.getInstance()
                try {
                    gpio = manager.openGpio(gpioName)
                    gpio.setDirection(Gpio.DIRECTION_OUT_INITIALLY_LOW)
                    gpio.setActiveType(Gpio.ACTIVE_HIGH)
                    gpio.value = true
                    gpioList.add(gpio)
                } catch (e: IOException) {
                    Log.e(TAG, "startRealPourThread: ", e)
                }
            }
            Log.d(TAG, "startRealPourThread start")
            delay(millis.toLong(), TimeUnit.MILLISECONDS)
            if (Utils.isThingsDevice(applicationContext)) {
                try {
                    gpio.setActiveType(Gpio.ACTIVE_LOW)
                    gpio.value = true
                    gpio.close()
                } catch (e: IOException) {
                    Log.e(TAG, "Error closing LED GPIO", e)
                }
            }
            Log.d(TAG, "startRealPourThread finish")
        }
        threadList.add(job)
//        val disposable = Completable
//                .fromAction {
//                    val manager: PeripheralManager
//                    if (Utils.isThingsDevice(applicationContext)) {
//                        manager = PeripheralManager.getInstance()
//                        try {
//                            gpio = manager.openGpio(gpioName)
//                            gpio.setDirection(Gpio.DIRECTION_OUT_INITIALLY_LOW)
//                            gpio.setActiveType(Gpio.ACTIVE_HIGH)
//                            gpio.value = true
//                        } catch (e: IOException) {
//                            Log.e(TAG, "startRealPourThread: ", e)
//                        }
//                    }
//                    Log.d(TAG, "startRealPourThread start")
//                }
//                .subscribeOn(Schedulers.io())
//                .delay(millis.toLong(), TimeUnit.MILLISECONDS)
//                .subscribe({
//                               if (Utils.isThingsDevice(applicationContext)) {
//                                   try {
//                                       gpio.setActiveType(Gpio.ACTIVE_LOW)
//                                       gpio.value = true
//                                       gpio.close()
//                                   } catch (e: IOException) {
//                                       Log.e(TAG, "Error closing LED GPIO", e)
//                                   }
//                               }
//                               Log.d(TAG, "startRealPourThread finish")
//                           }, { throwable ->
//                               Log.e(TAG, throwable.localizedMessage)
//                               Thread.currentThread().interrupt()
//                           })
//        threadList.add(disposable)
    }

    // TODO сделть возможность нливть по очереди

    private fun startProgressThread(step: Int, gpioId: Long) {
        val thread = Thread {

            //            val startTime = System.nanoTime()
            for (i in 0..100) {
                try {
                    Thread.sleep(step.toLong())
                    if (taskShouldBeStopped) {
                        Thread.currentThread().interrupt()
                        break
                    }
                } catch (e: InterruptedException) {
                    Log.e(TAG, "pourDemo: ", e)
                    Thread.currentThread().interrupt()
                }

                sendCocktailItemProgress(gpioId, i)
            }
//            val endTime = System.nanoTime()
//            val duration = endTime - startTime
//            Log.d(TAG, "startProgressThread: time $duration GPIO $gpioId")
        }
        thread.start()
    }

    private fun pour(motor: Motor, ml: Int, coefficient: Double) {
        val gpio = motor.gpio
        if (gpio != null) {
            pour(gpio.id, gpio.gpioName, motor.motorSpeed, ml, coefficient)
        } else {
            Log.e(TAG, "pour. gpio == null. motor = $motor")
        }
    }

    private fun pour(gpioId: Long, gpioName: String, motorSpeed: Int, ml: Int, coefficient: Double) {
        if (taskShouldBeStopped) {
            return
        }
        Log.d(TAG, "pour() called with: motor = [" + motor.toString() + "], ml = [" + ml + "]")
        val millis = ml * motorSpeed
        val step: Int = ((millis / 100) * coefficient).toInt()

        Log.d(TAG, "pour: millis = [$millis], step= [$step]")

        Log.d(TAG, "pour: start sleep")
        startProgressThread(step, gpioId)
        startRealPourThread(millis, gpioName)
    }

    companion object {

        const val BROADCAST_COCKTAIL_PROGRESS = "BROADCAST_COCKTAIL_PROGRESS"
        const val BROADCAST_COCKTAIL_STATUS = "BROADCAST_COCKTAIL_STATUS"
        const val BROADCAST_EXTRA_COCKTAIL_STATUS = "BROADCAST_EXTRA_COCKTAIL_STATUS"
        const val BROADCAST_EXTRA_COCKTAIL_ITEM_MAP = "BROADCAST_EXTRA_COCKTAIL_ITEM_MAP"
        private const val TAG = "CocktailService"
        private const val ACTION_COCKTAIL_CREATE = "ACTION_COCKTAIL_CREATE"
        private const val ACTION_COCKTAIL_CANCEL = "ACTION_COCKTAIL_CANCEL"
        private const val ACTION_MOTOR_TEST = "ACTION_MOTOR_TEST"
        private const val ACTION_MOTOR_CLEAR = "ACTION_MOTOR_CLEAR"
        private const val EXTRA_COCKTAIL_ID = "EXTRA_COCKTAIL"
        private const val EXTRA_MOTOR = "EXTRA_MOTOR"
        //        private const val EXTRA_MOTOR_ID = "EXTRA_MOTOR_ID"
//        private const val EXTRA_GPIO_ID = "EXTRA_GPIO_ID"
//        private const val EXTRA_GPIO_NAME = "EXTRA_GPIO_NAME"
        private const val EXTRA_GPIO_SPEED = "EXTRA_GPIO_SPEED"
        private const val EXTRA_MOTOR_LIST = "EXTRA_MOTOR_LIST"
        private const val EXTRA_MOTOR_CLEAR_TYPE = "EXTRA_MOTOR_CLEAR_TYPE"

        private var isTaskRun = false
        private var taskShouldBeStopped = false

        fun startCocktail(context: Context, cocktailId: Long) {
            val intent = Intent(context, CocktailService::class.java)
            intent.action = ACTION_COCKTAIL_CREATE
            intent.putExtra(EXTRA_COCKTAIL_ID, cocktailId)
            context.startService(intent)
        }

        fun stopCocktail(context: Context) {
            val intent = Intent(context, CocktailService::class.java)
            intent.action = ACTION_COCKTAIL_CANCEL
            context.startService(intent)
        }

        fun motorTest(context: Context, motor: Motor) {
            val intent = Intent(context, CocktailService::class.java)
            intent.action = ACTION_MOTOR_TEST
            intent.putExtra(EXTRA_MOTOR, motor)
//            intent.putExtra(EXTRA_GPIO_ID, motoror.gpioId)
//            intent.putExtra(EXTRA_GPIO_NAME, motor.gpio!!.gpioName)
            intent.putExtra(EXTRA_GPIO_SPEED, motor.motorSpeed)
            context.startService(intent)
        }


        fun clearShortMotors(context: Context, motorList: List<Motor>) {
            cleanMotors(context, motorList as java.util.ArrayList<Motor>, CleanType.SHORT)
        }

        fun clearLongMotors(context: Context, motorList: List<Motor>) {
            cleanMotors(context, motorList as java.util.ArrayList<Motor>, CleanType.LONG)
        }

        private fun cleanMotors(context: Context, motorList: ArrayList<Motor>, type: CleanType) {
            val intent = Intent(context, CocktailService::class.java)
            intent.action = ACTION_MOTOR_CLEAR
            intent.putExtra(EXTRA_MOTOR_LIST, motorList)
            intent.putExtra(EXTRA_MOTOR_CLEAR_TYPE, type)
            context.startService(intent)
        }
    }
}
